frugalos_config/src/cancelable_future.rs

```diff
@@ -0,0 +1,180 @@
+use futures::{Async, Future, Poll};
+use std::sync::mpsc;
+use {Error, ErrorKind};
+
+/// Signalを受信するための構造体であり、Futureとしても扱うことができる。  
+/// `make_signal`関数によって、対応するSenderとの組として生成される。
+pub struct SignalReceiver(mpsc::Receiver<()>);
+
+/// Signalを送信するための構造体である。  
+/// `make_signal`関数によって、対応するReceiverとの組として生成される。
+#[derive(Clone)]
+pub struct SignalSender(mpsc::Sender<()>);
+
+impl SignalSender {
+    /// 対応するReceiverにsignalを送信する。
+    pub fn send_signal(&mut self) -> Result<(), Error> {
+        self.0.send(()).map_err(Error::from)
+    }
+}
+
+impl SignalReceiver {
+    /// signalを取り出そうとする。
+    /// このメソッドは受信できるsignalがない場合にはエラーを返す。
+    pub fn try_recv_signal(&self) -> Result<(), Error> {
+        self.0.try_recv().map_err(Error::from)
+    }
+}
+
+impl Future for SignalReceiver {
+    type Item = ();
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let result = self.try_recv_signal();
+
+        match result {
+            Ok(()) => Ok(Async::Ready(())),
+            Err(e) => match e.kind() {
+                ErrorKind::WouldBlock => Ok(Async::NotReady),
+                _ => Err(e),
+            },
+        }
+    }
+}
+
+/// Signalの送信口`SignalSender`と受信口`SignalReceiver`を生成する。
+pub fn make_channel() -> (SignalSender, SignalReceiver) {
+    let (sender, receiver) = mpsc::channel();
+    (SignalSender(sender), SignalReceiver(receiver))
+}
+
+/// `future: Future<Item = (), Error = ()>`からキャンセル可能なFutureを作るための構造体である。
+/// # Example
+///
+/// ```ignore
+/// let mut executor = track!(ThreadPoolExecutor::new().map_err(Error::from))?;
+/// let mut rpc_server_builder = RpcServerBuilder::new(node.addr);
+/// ...
+/// let rpc_server = rpc_server_builder.finish(executor.handle());
+/// let (cancelable_rpc_server, cancelizer) =
+///     Cancelable::new(rpc_server.map_err(move |e| panic!("Error: {}", e)));
+/// ...
+/// executor.spawn(cancelable_rpc_server);
+/// ...
+/// // `some_future`は終了直前に、`rpc_server`を終させるために、`cancelizer.send_signal()`を呼ぶ。
+/// executor.run_future(some_future).unwrap();
+/// ```
+pub struct Cancelable {
```

Comment:
In my thought, combinator futures like this usually are defined as keeping the type signature of the internal future, e.g., `Cancelable<T>` (If a boxing is needed, users of the future will invoke `Box::new(Cancelable::new(...))` explicitly).

Reply:
Thank you. I try it!